pact/Pact/Core/TransitiveDependencies.hs

```diff
@@ -111,36 +170,85 @@ canHaveDependents = \case
 getAllTransitiveDependencies
-  :: (SizeOf b, SizeOf i)
-  => i
+  :: i
   -> S.Set ModuleName
   -> EvalModule b i
   -> EvalM e b i (M.Map FullyQualifiedName (EvalDef b i))
 getAllTransitiveDependencies i parentDirectDepSet mdl = isExecutionFlagSet FlagDisablePact51 >>= \case
   True -> oldDependencies
-  False -> allDependencies
+  False -> getAllDependencies
   where
-  -- When we traverse a term's dependents, if we don't see
-  -- a member already in the dependency set, then add it to the worklist and the dep set
-  computeTransitiveDeps worklist depSet = case popFromWorkingSet worklist of
-    -- We are done
-    Nothing -> return depSet
-    Just (e, rest) -> do
-      defn <- lookupFqNameOrFail i e
-      if canHaveDependents defn then do
-        defnSz <- sizeOf (defInfo defn) SizeOfV0 defn
-        chargeGasArgs i (GModuleOp (MOpFindTransitiveDep defnSz))
-        let (TransitiveClosureState newWorkList newDepSet) = execState (getDefDependents defn) (TransitiveClosureState rest depSet)
-        computeTransitiveDeps newWorkList newDepSet
-      else computeTransitiveDeps rest depSet
-
-  allDependencies = do
-    let initialDepSet =  S.fromList $ view _1 . toFqDep (_mName mdl) (_mHash mdl) <$> _mDefs mdl
-    DependencySet depSet <- computeTransitiveDeps (WorkingSet initialDepSet) (DependencySet initialDepSet)
-    let finalDepSet = depSet `S.difference` initialDepSet
-    (`M.restrictKeys` finalDepSet) <$> use (esLoaded . loAllLoaded)
+  tcloErrToPactError = \case
+    TCGasLimitExceeded mlim amt ->
+      PEExecutionError (GasExceeded (milliGasToGasLimit mlim) (milliGasToGas amt)) [] i
+    TCUnboundFreeVarError fqn -> PEExecutionError (InvariantFailure (InvariantUnboundFreeVariable fqn)) [] i
+
+  -- Calculate all of our dependencies from `computeTransitiveDeps`
+  getAllDependencies = do
+    let initialDepSet = S.fromList $ view _1 . toFqDep (_mName mdl) (_mHash mdl) <$> _mDefs mdl
+    allDeps <- use loAllLoaded
+    currGasConsumed@(MilliGas currGas) <- getGas
+    mgasLimit <- viewEvalEnv (eeGasEnv . geGasModel . gmGasLimit)
+    let tcloState = TransitiveClosureState
+                    { _workingSet = WorkingSet initialDepSet
+                    , _dependencySet = DependencySet initialDepSet
+                    , _gasConsumed = currGasConsumed
+                    , _gasLimit = mgasLimit
+                    , _allDependencies = allDeps }
+    case runState (runExceptT computeTransitiveDeps) tcloState of
+      (c, TransitiveClosureState _ws (DependencySet depSet) (MilliGas consumedByAlgorithm) _glim _allDeps) -> do
+        -- The way this module works, it starts at the current gas consumed, but does not charge gas
+        -- to our gas ref directly, so what we want is the difference of gas consumed post algorithm
+        -- exec vs gas pre-algorithm exec
+        -- We charge gas regardless here, which may raise
+        let consumedAmt = consumedByAlgorithm - currGas + restrictKeysCost allDeps depSet
```

Comment:
This is funny. We do almost the exact same thing for verifiers. Perhaps it should be a library function?

Reply:
Possibly, we might want a general interface for this?